### PR TITLE
elisp.nix: add a param to include user config as a default init file

### PR DESCRIPTION
**Pull request by jian-lin**
This is good for AOT native compilation of user config.

Besides being bool, the value of this new option can also be a derivation produced by pkgs.substituteAll, which avoids IFD comparing to directly setting that derivation to the config param.

cc @leungbk

### CODE REVIEW

**Comment by adisbladis** [comment]
Related issue comment: https://github.com/nix-community/emacs-overlay/issues/106#issuecomment-837102707

**Comment by jian-lin** [comment]
To be honest, I choose `default.el` just because it is what home-manager [does][1].

After reading @lambdadog's comparison of 4 methods, I still think `default.el` is the one to use in this overlay because it is less invasive, causes fewer problems and is easiest to implement. The only "problem" (I do not think it is a problem to me that it is loaded after `init.el`) is that it cannot handle the cases where `early-init.el` is needed. But that's not a big deal to me.

My initial goal is to do AOT native-comp for my emacs config file. I cannot make it work when specifying my config using home-manager's `extraConfig` together with `emacsWithPackagesFromUsePackage` like:
```nix
{
  programs.emacs = {
    enable = true;
    extraConfig = "...";
    package = pkgs.emacsWithPackagesFromUsePackage { ... };
  };
}
```
I guess it's because what home-manager thinks is emacs is actually a [wrapper][3] of emacs made by `emacsWithPackagesFromUsePackage`, and for that to work, a horrible hack can be to change `lib.optionalAttrs (emacs.nativeComp or false)` to `lib.optionalAttrs (emacs.emacs.nativeComp or false)` in [generic.nix][2]. So I choose to implement this pr, which is more elegant in my opinion.

Please note that, currently my implementation does not support org files and I do not have strong interest to support it since I do not use it myself. I suppose it's fine to leave that for future contributors and merge this pr for now. But if that support is needed to get this merged, please let me know and I might try to implement that, not a priority for me though.

[1]: https://github.com/nix-community/home-manager/blob/ebe6d2c747cc6e2223dd5962bdca258088518b3f/modules/programs/emacs.nix#L19
[2]: https://github.com/NixOS/nixpkgs/blob/ef0ed284ea6ce8d14de9575d1962e70590c29bfe/pkgs/build-support/emacs/generic.nix#L71
[3]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/emacs/wrapper.nix

File: elisp.nix [hunk 1]
```diff
@@ -64,5 +66,28 @@ emacsWithPackages (epkgs:
     overridden = override epkgs;
     usePkgs = map (name: overridden.${name} or (mkPackageError name)) packages;
     extraPkgs = extraEmacsPackages overridden;
+    defaultInitFilePkg =
+      let
+        isBool = builtins.isBool defaultInitFile;
+        isDerivation = lib.isDerivation defaultInitFile;
+      in
+      if !(isBool || isDerivation)
+      then throw "defaultInitFile must be bool of derivation"
```
Comment (leungbk) [inline elisp.nix:75]:
```suggestion
      then throw "defaultInitFile must be bool or derivation"
```

**Comment by leungbk** [comment]
Thanks!

**Pull request merged**